File: web/src/integration/titaniumdrm/TitaniumWidevineContentProtectionIntegration.ts
import { isTitaniumDRMConfiguration } from './TitaniumUtils';
import { createTitaniumHeaders, TitaniumCDMType } from './TitaniumBaseRegistration';
import type { BufferSource, ContentProtectionIntegration, LicenseRequest, MaybeAsync } from 'THEOplayer';
import type { TitaniumDrmConfiguration } from './TitaniumDrmConfiguration';
import type { CertificateRequest } from 'THEOplayer';

export class TitaniumWidevineContentProtectionIntegration implements ContentProtectionIntegration {
    private readonly contentProtectionConfiguration: TitaniumDrmConfiguration;

    constructor(drmConfiguration: TitaniumDrmConfiguration) {
        if (!isTitaniumDRMConfiguration(drmConfiguration)) {
            throw new Error('Titanium DRM has not been correctly configured.');
        }
        this.contentProtectionConfiguration = drmConfiguration;
    }

    onCertificateRequest(request: CertificateRequest): MaybeAsync<Partial<CertificateRequest> | BufferSource> {
        request.headers = {
            ...request.headers,
            ...createTitaniumHeaders(this.contentProtectionConfiguration, TitaniumCDMType.WIDEVINE),
        };
        return request;
    }

    onLicenseRequest(request: LicenseRequest): MaybeAsync<Partial<LicenseRequest> | BufferSource> {
        request.headers = {
            ...request.headers,
            ...createTitaniumHeaders(this.contentProtectionConfiguration, TitaniumCDMType.WIDEVINE),
        };
        return request;
    }
}

File: android/app/src/main/java/com/theoplayer/contentprotectionintegration/integration/titanium/TitaniumWidevineContentProtectionIntegration.kt
package com.theoplayer.contentprotectionintegration.integration.titanium

import com.theoplayer.android.api.contentprotection.*
import com.theoplayer.android.api.source.drm.DRMConfiguration
import com.theoplayer.contentprotectionintegration.integration.titanium.TitaniumBaseRegistration.createTitaniumHeaders

class TitaniumWidevineContentProtectionIntegration(private val configuration: DRMConfiguration):
    ContentProtectionIntegration() {

    override fun onCertificateRequest(request: Request, callback: CertificateRequestCallback) {
        request.apply {
            headers.putAll(createTitaniumHeaders(configuration, TitaniumWidevineCDMDescription))
            callback.request(this)
        }
    }

    override fun onLicenseRequest(request: Request, callback: LicenseRequestCallback) {
        request.apply {
            headers.putAll(createTitaniumHeaders(configuration, TitaniumWidevineCDMDescription))
            callback.request(this)
        }
    }
}
